File: MatrixFormulation/Python/MatrixFormulation.py
import numpy as np
import networkx as nx
import csv
import sys

file="../Graph/directedEdges.csv"

beta=0.9

# create networkx graph
G=nx.DiGraph()

#Open File and read directed Edges from each line
with open(file) as f:
    reader = csv.reader(f)
    for row in reader:
       G.add_edge(row[0], row[1])

graphNodes = G.nodes()
importanceContributionDictionary = {}

for gnode in graphNodes:
   #calculate outdegree for each node
   nodeOutDegree = G.out_degree(gnode)
   #if outdegree is 0, importance contribution is 0 else 1/outdegree
   if nodeOutDegree == 0 :
      importanceContributionDictionary[gnode] = 0.0
   else:
      importanceContributionDictionary[gnode] = 1.0/nodeOutDegree

#length of graphNodes
graphNodesLength = len(graphNodes)

#Initialize stochastic matrix with zeroes
stochasticMatrix = np.zeros((graphNodesLength,graphNodesLength))

i = 0
#replace zeroes with coefficients in stochasticMatrix
for gnode in graphNodes:
   nodeInnerEdges = G.in_edges(gnode)
   for inLinkEdgeNode in [linkEdge[0] for linkEdge in nodeInnerEdges]:
      stochasticMatrix[i,graphNodes.index(inLinkEdgeNode)] = 1.0*importanceContributionDictionary[inLinkEdgeNode]
   i = i + 1

#add (1-beta)/numberOfNodes*matrixwithallones for spidertraps, deadends and self-loops
stochasticMatrix = beta*stochasticMatrix + (1-beta)/graphNodesLength*np.ones((graphNodesLength,graphNodesLength))

#initialize random page ranks for each node
pageRanks = np.random.random_sample((graphNodesLength,))

#sum of pageranks must be one
pageRanks = pageRanks/pageRanks.sum()


#Iterative Calculation for Pageranks
for i in range(0, 1000):
     tempPageRanks = np.dot(stochasticMatrix,pageRanks)
     tempPageRanks = tempPageRanks/tempPageRanks.sum()
     if np.absolute(pageRanks - tempPageRanks).sum() < 0.000001:
        pageRanks = tempPageRanks
        break
     else:
        pageRanks = tempPageRanks

#zip with graphNodes to know pageRanks of each corresponding node
dictionary = dict(zip(graphNodes, pageRanks))

#print node and its page rank
print dictionary

File: MatrixFormulation/R/MatrixFormulation.R
#load igraph package for graph analysis
require(igraph)

beta = 0.9
#load directed edges from the file
directedEdges <- read.csv("../Graph/directedEdges.csv", header=FALSE)

#convert above dataframe into graph data frame of igraph
directedEdges.network<-graph.data.frame(directedEdges, directed=T)

#Calculate number of vertices from the graph
noofvertices <- length(V(directedEdges.network)$name)

#initialize stochastic matrix with zeroes
stochasticMatrix <- matrix(0,noofvertices,noofvertices)

#Creating Coefficient matrix from the data
i = 1
for (node in sort(V(directedEdges.network)$name)) {
  for (vertex in V(directedEdges.network)[nei(node,"in")]){
    stochasticMatrix[i,match(vertex,V(directedEdges.network)$name)] = 1.0/as.integer(degree(directedEdges.network,vertex,mode=c("out")))
  }
  i = i+1
}

#add (1-beta)/numberOfNodes*matrixwithallones for spidertraps, deadends and self-loops
stochasticMatrix = beta*stochasticMatrix + (1-beta)/noofvertices*matrix(1,noofvertices,noofvertices)

#initialize random page ranks for each node
pageRanks = rep(1.0/noofvertices,noofvertices)

#sum of pageranks must be one
pageRanks = pageRanks/sum(pageRanks)

#Iterative Calculation for Pageranks
for (i in 1:1000){
#  print(pageRanks)
  tempPageRanks <- stochasticMatrix %*% pageRanks
#  print(tempPageRanks)
  rankSum = sum(tempPageRanks)
  tempPageRanks = tempPageRanks + (1-rankSum)/noofvertices
  if (sum(abs(pageRanks - tempPageRanks)) < 0.000001){
    pageRanks = tempPageRanks
    break
  }else{
   pageRanks = tempPageRanks
  }
}

#Printing Vertices First
print(sort(V(directedEdges.network)$name))
#Printing PageRanks of above vertices
print(pageRanks)
